Support requesting resources from the search bar

gzdunek opened the pull request:
In addition to requesting resources from the unified view, we can do this also from the search bar.

<img width="1272" alt="image" src="https://github.com/gravitational/teleport/assets/20583051/6e28b62c-cd2d-405b-8081-baf031ed068c">

To keep the search bar simple, I always show `Request access to...` for requestable resources, even if they were already added to the request. I didn't want to make the search bar more complex with additional states.

gzdunek committed:
Ask for requestable resources in the search bar if the cluster allows showing them

gzdunek committed:
Show requestable resources in the list

gzdunek committed:
Add a resource to the request when the user selects it

gzdunek commented on web/packages/teleterm/src/ui/Search/useSearch.ts:
```diff
@@ -360,6 +364,11 @@ function calculateScore(
 
+  // Show resources that require access at the bottom.
+  if (searchResult.requiresRequest) {
+    searchResultScore *= 0.1;
```
This doesn't prevent the normal resource from having a score lower than a requestable one, but perhaps it is good enough?

ravicious replied:
What's the reasoning behind such a severe penalty though? I think something like `*= 0.95` should be good enough for now.

Imagine this scenario:

```diff
diff --git a/web/packages/teleterm/src/ui/Search/useSearch.test.tsx b/web/packages/teleterm/src/ui/Search/useSearch.test.tsx
index 5d7e76d6ff..2f87de779b 100644
--- a/web/packages/teleterm/src/ui/Search/useSearch.test.tsx
+++ b/web/packages/teleterm/src/ui/Search/useSearch.test.tsx
@@ -54,6 +54,33 @@ describe('rankResults', () => {
     expect(sortedResults[1]).toEqual(server);
   });
 
+  it('prefers accessible resources over requestable ones', () => {
+    const serverAccessible = makeResourceResult({
+      kind: 'server',
+      resource: makeServer({ hostname: 'sales-foo' }),
+    });
+    const serverRequestable = makeResourceResult({
+      kind: 'server',
+      resource: makeServer({ hostname: 'sales-bar' }),
+      requiresRequest: true,
+    });
+    const labelMatch = makeResourceResult({
+      kind: 'server',
+      resource: makeServer({
+        hostname: 'lorem-ipsum',
+        labels: makeLabelsList({ foo: 'sales' }),
+      }),
+    });
+    const sortedResults = rankResults(
+      [labelMatch, serverRequestable, serverAccessible],
+      'sales'
+    );
+
+    expect(sortedResults[0].resource).toEqual(serverAccessible.resource);
+    expect(sortedResults[1].resource).toEqual(serverRequestable.resource);
+    expect(sortedResults[2].resource).toEqual(labelMatch.resource);
+  });
+
   it('saves individual label match scores', () => {
     const server = makeResourceResult({
       kind: 'server',

```

The scores are as follows:

```
yours (=* 0.1):
[ 'sales-foo: 220', 'lorem-ipsum: 100', 'sales-bar: 22' ]

mine (=* 0.95):
[ 'sales-foo: 220', 'sales-bar: 209', 'lorem-ipsum: 100' ]
```

In this scenario, I don't believe we'd want a resource with a single hit on a label to be above a requestable resource with a hit on its name.

gzdunek replied:
Ok, I could be too aggressive on that 😅 My first instinct was that requestable resources should be always displayed at the bottom, but yeah, just having a lower priority may be fine.

Changed to 0.95, I also added the test, thanks!

gzdunek commented on web/packages/teleterm/src/ui/Search/pickers/ActionPicker.tsx:
```diff
@@ -608,15 +608,21 @@ export function ServerItem(props: SearchResultItem<SearchResultServer>) {
         <Text typography="body1">
-          Connect over SSH to{' '}
+          {props.searchResult.requiresRequest
+            ? 'Request access to server '
```
In the unified view this works slightly different: if you already have a resource in the access request, all buttons (even on accessible resources) change to 'Request access'. This is because when you assume a resource request, you lose access to other things that aren't in the request. 

However, in the search bar I only show 'Request access' for the resources that the user doesn't have access to. What are your thoughts on this? Should the search bar switch to the 'request resources' mode too (for a given workspace)? cc @avatus

ravicious replied:
One argument against switching the search bar to the request resources mode would be that access requests are scoped per root cluster. So the user would end up with some resources being in "request access" mode and others in "connect" mode.

This is not necessarily bad and having multiple root clusters is rare. Nonetheless, it's one of the use cases we built the search bar for.

I don't have a strong feeling on this, but as you said in the PR description, I think we should keep the search bar as simple as possible. This is the same reason for which for TCP apps you cannot _either_ use VNet or a local proxy from the search bar. You can only select one action which is the preferred one.

avatus replied:
I agree with both of your sentiments here. IMO, no one is going to be "building" an access request through the search bar, but only starting one. The rest is most likely done in the document. (of course, this is an assumption). I think the search bar as simple as possible is 👍

gzdunek replied:
Thanks for the feedback guys, looks like we all agree on keeping the search bar simple!

ravicious commented on web/packages/teleterm/src/ui/services/workspacesService/accessRequestsService.ts:
Not strictly related to this comment, but in the Web UI clicking "Cancel" in the access request sidebar removes the request completely, but in Connect it's an equivalent of clicking the back arrow next to the sidebar title.

Was it intentional? Without such button, requesting a resource switches the UI to a new mode (where you add resources to a request rather than connect to them), but there's no easy way to exit that mode.

gzdunek replied:
Right, these two elements in the UI were doing the same action, I aligned it with Web UI.

ravicious commented on web/packages/teleterm/src/ui/services/resources/resourcesService.ts:
```diff
@@ -124,11 +126,11 @@ export class ResourcesService {
         query: '',
-        searchAsRoles: false,
+        searchAsRoles: includeRequestable,
         pinnedOnly: false,
         startKey: '',
         sortBy: { field: 'name', isDesc: true },
-        includeRequestable: false, //TODO(gzdunek): Support requestable resources in the search bar.
+        includeRequestable,
```
Is there ever a situation where `searchAsRoles` and `includeRequestable` in the RPC won't have the same value?

avatus replied:
I believe we are going to be deprecating `searchAsRoles` in the future. `includeRequestable` handles all the cases we need (accessible, requestable, all) but I haven't added in the distinction yet. You might be able to get away with not adding `searchAsRoles` to this call at all now because of https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L2660

and
https://github.com/gravitational/teleport/blob/master/lib/auth/auth_with_roles.go#L1328

but i haven't looked at the tsh handler to confirm this

gzdunek replied:
Yeah, having these two options at the same time is confusing and we talked about deprecating it. 
Unfortunately, it may require adding another parameter instead of `includeRequestable` which is only a boolean. 
I think you Michael meant `includedResourceMode` (that has three values: accessible, requestable, all) which lives only in the web api layer.

ravicious commented on web/packages/teleterm/src/ui/Search/actions.tsx:
```diff
@@ -261,3 +294,17 @@ export function mapToAction(
   }
 }
+
+async function addResourceToRequest(
+  ctx: IAppContext,
+  resource: ResourceRequest
+): Promise<void> {
+  const rootClusterUri = routing.ensureRootClusterUri(resource.resource.uri);
+  const { isAtDesiredWorkspace } =
+    await ctx.workspacesService.setActiveWorkspace(rootClusterUri);
+  if (isAtDesiredWorkspace) {
+    await ctx.workspacesService
+      .getWorkspaceAccessRequestsService(rootClusterUri)
+      .addResource(resource);
+  }
+}
```
Selecting an resource I've already added to a request sends a `GetRequestableRoles` RPC. Is there a simple way in which we could avoid that?

avatus replied:
At the moment, resources exist in a state of "you have access to it" and "requires request". Off the top of my head, we could maybe avoid the fetch if we do some sort of check like "if this is being added to the request but `requiresRequest` isn't explicitly `true`" (which would imply we are in "request" mode), then maybe don't fetch it. however, that could interfere with the roles required for the two resources to be available (assuming some weird interaction with roles needed for the explicitly `requiresRequest` resources, and the ones you have access to already).

Long story longer, I believe we should just call it all times to get the "most correct" list. I won't die on this hill tho

gzdunek replied:
Actually, the reason is quite simple, when adding a resource again, we overwrite it and the state changes. This in turn causes `useEffect` to fire the request. 
I added a check that skips adding a resource if it is already added.

ravicious reviewed (commented):
It looks good overall, submitted some minor suggestions.

gzdunek committed:
Do not a resource if it is already added

gzdunek committed:
Increase score for requestable resources, add a test

gzdunek committed:
Cancel button should clear the request

gzdunek committed:
Do not set `searchAsRoles`

gzdunek committed:
Reset the unified resources component when `rootCluster.showResources` changes

I noticed that sometimes the component stuck on 'No resources' screen.
It happens when the call for unified resources is done before cluster details are fetched.

ravicious commented on web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx:
```diff
@@ -230,6 +230,10 @@ const Resources = memo(
     const { fetch, resources, attempt, clear } = useUnifiedResourcesFetch({
       fetchFunc: useCallback(
         async (paginationParams, signal) => {
+          // Do not make the call if we don't know what resources we should show.
+          if (props.showResources === ShowResources.UNSPECIFIED) {
+            return { startKey: '', agents: [] };
+          }
```
How does this work? Just by looking at this part of the code alone, it appears that it's wrong: if we get to this point before receiving cluster details, then the resources tab is going to flash "No resources" for a brief time before actually making a request.

To test this, I made the `GetCluster` RPC artificially slower by 3s on a local cluster. It seems to be handled correctly, the tab is stuck on the loading spinner.

<image alt="but why" src="https://storage.googleapis.com/sourcegraph-assets/blog/ryan-reynold-but-why-meme.jpg" width="250"/>

ravicious replied:
Oh, I see, it's because of this:

https://github.com/gravitational/teleport/blob/f2907754c2630698617315b783269a9dfe0a6dcd/web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx#L325-L329

ravicious replied:
I wonder if it wouldn't make more sense to make `fetchFunc` block forever until `showResources` are fetched? Otherwise the code that does the fetching is desynchronized with the code that reports the status of the fetching. This is what led to my original comment here in the first place.

We could have a function similar to wait, but it'd only take a signal and never resolve until the signal is aborted.

gzdunek replied:
In the first iteration I had something like `waitForever` but then I replaced it with that combined `resourcesFetchAttempt`. 
I'm happy to bring it back, it will make the code easier to understand. I had to adjust the code `useKeyBasedPagination.ts` so now it aborts the pending request when the hook unmounts.

gzdunek committed:
Block the call for `listUnifiedResources` when `showResources` is unspecified

ravicious commented on web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx:
```diff
@@ -512,3 +511,17 @@ function NoResources(props: {
   );
 }
+
+function waitForever(abortSignal: AbortSignal): Promise<never> {
```
This could go straight to `utils/wait.ts`, no? ;)

ravicious commented on web/packages/teleterm/src/ui/DocumentCluster/UnifiedResources.tsx:
```diff
@@ -230,9 +230,10 @@ const Resources = memo(
+          // We will remount the component and do the call when showResources changes.
           if (props.showResources === ShowResources.UNSPECIFIED) {
-            return { startKey: '', agents: [] };
+            await waitForever(signal);
```
Nice, good comment too. 👍

gzdunek committed:
Move `waitForever` to `shared/utils`
